jobs,sql: create jobs in a batch in a transactions

sajjadrizvi opened the pull request:
This commit change the way jobs are created for a transaction.
Previously, the jobs were created individually during the
execution of a transaction. A job creation incurs a round trip
to insert an entry in system.jobs table. 

This commit introduces a new function in Registry, 'CreateJobsWithTxn',
that allows creating jobs in a batch using a single SQL statement.  Using
this function, this commit creates jobs in a batch during a transaction, which
reduces the number of round-trips between distant nodes during a transaction
execution. This is a step forward to improve the performance of GRANT/REVOKE 
queries in large databases.

Release note: None

Fixes: #64389

Commit:
jobs: create jobs in a batch

This commit adds a new function in Registry, 'CreateJobsWithTxn',
that allows creating jobs in a batch using a single SQL statement.
This function is a precursor to creating schema-changing jobs
in a batch, which will result in reducing the number of round-trips
between nodes during schema-changes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67991)
<!-- Reviewable:end -->

Commit:
sql: create jobs in a batch in a transactions

This commit changes the way schema-change jobs are created in a
transaction. Previously, the jobs were created individually during the
execution of a transaction. A job creation incurs a round trip to
insert an entry in jobs table. This commit accumulates the jobs
to be created during a transaction and creates them in a batch, which
is expected to reduce the number of round-trips between distant nodes
during a transaction execution. This is a step forward to improve the
performance of GRANT/REVOKE queries in large databases.

Release note: None

Fixes: #64389

sajjadrizvi commented:
Although some tests are failing, I appreciate if I can get some feedback. I might be pushing too hard in the tests that are failing. If not, they may lead to further changes or adding bounds on the size of a batch.

ajwerner commented on pkg/jobs/jobs.go:
```diff
@@ -76,6 +76,50 @@ type Record struct {
 
+// Specification specifies a job, which will be created in a batch with other jobs.
+type Specification struct {
+	jobID  jobspb.JobID
```
I'm starting to think maybe we should just stick ID as a field in Record. It'd simplify the code and some function signatures. I don't see any good reason not to. The history here is that the jobs package used to allocate the id for the user. That turned out to be unfortunate so we changed the client to allocate the id. May as well just move the ID into the record.

ajwerner commented on pkg/jobs/registry.go:
```diff
@@ -368,6 +368,73 @@ func (r *Registry) newJob(record Record, jobID jobspb.JobID) *Job {
+// in a single insert to the jobs table.
+// There must be at least one job to create, otherwise the function returns an error.
+// If an error occurs while preparing the insert statement, no job is created.
+//TODO(sajjad): To discuss: What batch size do we expect? Should we limit the
```
Yeah we'll want a maximum batch size. Start with just some constant. Maybe 100? We'll want to make sure we test using multiple batches.

ajwerner commented on pkg/jobs/registry.go:
```diff
@@ -368,6 +368,73 @@ func (r *Registry) newJob(record Record, jobID jobspb.JobID) *Job {
+// column's value will be large due to large statement size. Are there any
+// implications that can result in exceeding limits such as columns size limits?
+func (r *Registry) CreateJobsWithTxn(
+	ctx context.Context, jobSpecs []*Specification, txn *kv.Txn,
```
nit: put the transaction before the jobs

ajwerner commented on pkg/sql/planner.go:
```diff
@@ -93,7 +93,7 @@ type extendedEvalContext struct {
 
 	// SchemaChangeJobCache refers to schemaChangeJobsCache in extraTxnState.
-	SchemaChangeJobCache map[descpb.ID]*jobs.Job
+	SchemaChangeJobCache map[descpb.ID]*jobs.Specification
```
My guess is that we're going to want to nuance this data structure a bit. There are places where we delete from this cache. Those places are trying to be a bit clever and used to rely on the fact that we had already written the job. I'm thinking what we want is to create a slice of job records and a map from descriptor ID to the record. Then in the places where today we delete from the map we'd retain the record in the slice. 

I suspect this is the cause of some of the failing tests.

ajwerner commented on pkg/sql/schema.go:
```diff
@@ -83,11 +83,11 @@ func (p *planner) writeSchemaDescChange(
-		}
-		log.Infof(ctx, "queued new schema change job %d for schema %d", newJob.ID(), desc.ID)
+		newSpec := jobs.NewSpecification(p.extendedEvalCtx.ExecCfg.JobRegistry.MakeJobID(), jobRecord)
+		//TODO(sajjad): To discuss: Previous code didn't cache the new job. I think
```
Can you look at what QueueJob does?

ajwerner reviewed (commented):
Preliminary pass on the subway back from the dentist

sajjadrizvi reviewed (commented):
Thanks for the quick review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-MfY_cCf9dh4T7nFjGi2:b-ncgzbp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/jobs/jobs.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g1-r2-81:-MfYhKKG4Fs9Me8NpAaO:bchvim9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/jobs.go#L81)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm starting to think maybe we should just stick ID as a field in Record. It'd simplify the code and some function signatures. I don't see any good reason not to. The history here is that the jobs package used to allocate the id for the user. That turned out to be unfortunate so we changed the client to allocate the id. May as well just move the ID into the record. 
</blockquote></details>

I have not built a sufficient context yet. I am following your judgement here.
___
*[pkg/jobs/registry.go, line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g2-r2-375:-MfYhU5IAs9hCfXgwpvX:b-b4n8p7) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/registry.go#L375)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Yeah we'll want a maximum batch size. Start with just some constant. Maybe 100? We'll want to make sure we test using multiple batches. 
</blockquote></details>

I will set a max batch size. We may want to make it a cluster setting, but not for now.

I already have a test case that checks multiple batches. The test fails for larger batches.
___
*[pkg/jobs/registry.go, line 380 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g2-r2-380:-MfZ3H1_1nK1j9-gX3C_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/registry.go#L380)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

nit: put the transaction before the jobs
</blockquote></details>

Done.
___
*[pkg/sql/planner.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfN1ZkROsZJINEQBmxO-r2-96:-MfZ3MVPDscJSTcBfeyM:bsznc7f) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/planner.go#L96)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

My guess is that we're going to want to nuance this data structure a bit. There are places where we delete from this cache. Those places are trying to be a bit clever and used to rely on the fact that we had already written the job. I'm thinking what we want is to create a slice of job records and a map from descriptor ID to the record. Then in the places where today we delete from the map we'd retain the record in the slice. 

I suspect this is the cause of some of the failing tests. 
</blockquote></details>

OK. I guess that clever function is `markTableMutationJobsSuccessful`. We will end up deleting the slice only in`resetExtraTxnState`.

It might be the cause of error! However, the test logs show messages related to throttling and ranges going down.
___
*[pkg/sql/schema.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfN1ZkROsZJINEQBmxP-r2-87:-MfZ7von7_5AITNNi_4P:b-l7tjjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/schema.go#L87)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Can you look at what QueueJob does?
</blockquote></details>

Yeah, it creates a new job and adds the job to Jobs collection. So this is a place where want the job to be only in the Jobs collection but not in the cache. I will have to preserve this. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sajjadrizvi)

___
*[pkg/jobs/jobs.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g1-r2-81:-MfZB96jC4-OzxKTi_-X:b-skyodk) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/jobs.go#L81)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

I have not built a sufficient context yet. I am following your judgement here.
</blockquote></details>

Yeah, my advice is that we add `ID` as a field to record. It should simplify the interfaces. 
___
*[pkg/jobs/registry.go, line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g2-r2-375:-MfZB4Bh2RvjW2yGXzjN:b9xqsk1) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/registry.go#L375)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

I will set a max batch size. We may want to make it a cluster setting, but not for now.

I already have a test case that checks multiple batches. The test fails for larger batches.
</blockquote></details>

It should not fail. Instead this function should have a loop to send multiple batches. 
___
*[pkg/sql/conn_executor_exec.go, line 807 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZ9Uqv0iewL5EMSiPR:-MfZ9Uqw7wTIGN65VPrc:b-xzmube) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/conn_executor_exec.go#L807)):*
<details><summary><i alt="summary">Quoted 10 lines of code…</i></summary>

> ```
> 
> 	// 1. I am not sure if I am returning the error correctly.
> 	// 2. Is this the right place to create jobs? I have placed it outside of the
> 	//    the following code block because the jobs were being created outside of
> 	//    it in the previous code.
> 	// 3. If I have add any ex.phaseTimes, which one should it be? Do I have to
> 	//    create a new one?
> 	if err := ex.createJobs(ctx); err != nil {
> 		return ex.makeErrEvent(err, ast)
> 	}
> ```
</details>

This logic should be pulled into `commitSQLTransactionInternal`
___
*[pkg/sql/conn_executor_exec.go, line 824 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZB0_z38LZtsk11noz:-MfZB0_z38LZtsk11np-:bc0gw0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/conn_executor_exec.go#L824)):*
> ```Go
> 	}
> 	var specList []*jobs.Specification
> 	//TODO(sajjad): To discuss: Does the order of jobs creation matter?
> ```

No, it does not matter. 
___
*[pkg/sql/planner.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfN1ZkROsZJINEQBmxO-r2-96:-MfZAtH5-fGVGFzvziN4:bl4a6ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/planner.go#L96)):*
> the test logs show messages related to throttling and ranges going down.

I have a feeling that that's just noise at the end of the test and is not the true failure. 
___
*[pkg/sql/schema.go, line 87 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfN1ZkROsZJINEQBmxP-r2-87:-MfZAUukAnwaxntBAmyu:bblct1x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/schema.go#L87)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

Yeah, it creates a new job and adds the job to Jobs collection. So this is a place where want the job to be only in the Jobs collection but not in the cache. I will have to preserve this. 
</blockquote></details>

My guess is that not putting the descriptor into the cache was an oversight. 
___
*[pkg/sql/schema.go, line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZAYlZC7RZDIK4Rzbw:-MfZAYlZC7RZDIK4Rzbx:b-heopn6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/schema.go#L90)):*
> ```Go
> 		// the newSpec should be cached.
> 		p.extendedEvalCtx.SchemaChangeJobCache[desc.ID] = newSpec
> 		log.Infof(ctx, "cached specification of new schema change job %d for schema %d", newSpec.ID(), desc.ID)
> ```

This is not a good log message. It would be extremely confusing for a read. I think the old message was pretty reasonable. Your change of the implementation here doesn't affect the behavior really at all. 


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/registry.go, line 375 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g2-r2-375:-Mf_u9NwC6jRRi_ctSQs:b-cc00ff) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/registry.go#L375)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

It should not fail. Instead this function should have a loop to send multiple batches. 
</blockquote></details>

I meant the batch size set in the test function. Nevertheless, I have implemented as suggested.
___
*[pkg/sql/conn_executor_exec.go, line 807 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZ9Uqv0iewL5EMSiPR:-Mfa4NE57xqTblcKmRf2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/conn_executor_exec.go#L807)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 	// 1. I am not sure if I am returning the error correctly.
> 	// 2. Is this the right place to create jobs? I have placed it outside of the
> 	//    the following code block because the jobs were being created outside of
> 	//    it in the previous code.
> 	// 3. If I have add any ex.phaseTimes, which one should it be? Do I have to
> 	//    create a new one?
> 	if err := ex.createJobs(ctx); err != nil {
> 		return ex.makeErrEvent(err, ast)
> 	}
> ```
This logic should be pulled into `commitSQLTransactionInternal`
</blockquote></details>

Done.
___
*[pkg/sql/conn_executor_exec.go, line 824 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZB0_z38LZtsk11noz:-Mfa4S7B6upqMAAOHRdq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/conn_executor_exec.go#L824)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

No, it does not matter. 
</blockquote></details>

Done.
___
*[pkg/sql/schema.go, line 90 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfZAYlZC7RZDIK4Rzbw:-MfaHn9YBxig0745YRYr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/schema.go#L90)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This is not a good log message. It would be extremely confusing for a read. I think the old message was pretty reasonable. Your change of the implementation here doesn't affect the behavior really at all. 
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
Sorry I forgot to publish the comments before pushing the changes. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-MfaRN4M6TGVTh5A0U_H:bxozklu)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)



<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
This PR is now ready for another review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-MfcEVyk3PHQAmIoLUYC:b-y2p5ne)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Nice, this is getting close. 

> Reviewed 5 of 14 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-Mfcx2d932K4ttDaW0JT:b-rn7qjg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @sajjadrizvi)

___
*[pkg/bench/rttanalysis/testdata/benchmark_expectations, line 80 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcknDe6Bur1HJ6OGHh:-MfcknDfCHjVhHxx-ol2:b-ty67fc) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/bench/rttanalysis/testdata/benchmark_expectations#L80)):*
<details><summary><i alt="summary">Quoted 6 lines of code…</i></summary>

> ```
> 26,Truncate/truncate_1_column_0_rows
> 26,Truncate/truncate_1_column_1_row
> 26,Truncate/truncate_1_column_2_rows
> 26,Truncate/truncate_2_column_0_rows
> 26,Truncate/truncate_2_column_1_rows
> 26,Truncate/truncate_2_column_2_rows
> ```
</details>

I have a feeling we're going to want to retain the range here of round-trips here (put the 26-27 back). 
___
*[pkg/jobs/registry.go, line 375 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfctsoy8qg5KjUp9whS:-Mfctsoy8qg5KjUp9whT:b-r231hb) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L375)):*
> no further jobs are created and the error is returned.

I find this language confusing. The caller shouldn't need to care about other jobs having been created. It makes it sound like it is not a transactional operation. 
___
*[pkg/jobs/registry.go, line 378 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcwu1wDAApIJ8Yg4cb:-Mfcwu1wDAApIJ8Yg4cc:b-33vl4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L378)):*
> ```Go
> func (r *Registry) CreateJobsWithTxn(
> 	ctx context.Context, txn *kv.Txn, records []*Record,
> ) ([]*Job, error) {
> ```

Do callers want the `*Job` or would they be happy with just a slice of IDs? I feel like IDs might be better.
___
*[pkg/jobs/registry.go, line 382 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcu92e1dLlDiHL3z72:-Mfcu92e1dLlDiHL3z73:bdwrb4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L382)):*
> ```Go
> 		return nil, errors.Errorf("no jobs to create.")
> 	}
> ```

can you refactor this a bit into a helper to write a batch and a helper to generate the statement and args. I think it should read something like:

```
created := make([]*Job, 0, len(records)
for toCreate := records; len(toCreate) > 0; {
    const maxBatchSize = 100
    batchSize := len(toCreate)
    if batchSize > maxBatchSize {
        batchSize = maxBatchSize
    }
    createdInBatch, err := r.createJobsWithTxnInBatch(ctx, toCreate[:batchSize])
    if err != nil {
        return nil, err
    }
    created = append(created, createdInBatch...)
    toCreate = toCreate[batchSize:]
}

func (r *Registry) createJobsWithTxnInBatch(ctx context.Context, txn *kv.Txn records []*Record) ([]*Job, error) {
    s, err  := r.sqlInstance.Session(ctx)
    if err != nil {
        return nil, errors.Wrap(err, "error getting live session")
    }
    stmt, args := batchJobInsertStmt(s, records)
    if err := r.ex.Exec(ctx, "job-rows-batch-insert", txn, stmt, args...); err != nil {
        return nil, err
    }
    // return the jobs etc. Maybe construct them when you construct the statement and args. 
}
```
___
*[pkg/jobs/registry.go, line 454 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcl1tF220t_nj1_YFo:-Mfcl1tGB4_PegMQz9-u:bsrxgdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L454)):*
> ```
> 			if _, err = r.ex.Exec(ctx, "job-rows-batch-insert", txn, stmtSB.String(), args...,
> 			); err != nil {
> ```
nit: I find this wrapping to be bizarre. My preference is:

```go
		if _, err = r.ex.Exec(
			ctx, "job-rows-batch-insert", txn, stmtSB.String(), args...,
		); err != nil {
```
___
*[pkg/jobs/registry_test.go, line 302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcrZaJ-SOdN04NLpSY:-MfcrZaK2rYqAcUfZgjf:b-m5z4x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L302)):*
> ```Go
> 				}
> 
> 				skip.UnderStress(t)
> ```

Isn't this covered above?
___
*[pkg/jobs/registry_test.go, line 334 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcrg254D2oycJJylk0:-Mfcrg254D2oycJJylk1:b-l2x18s) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L334)):*
> ```Go
> 				// Create jobs in a batch.
> 				var jobs []*Job
> 				require.NoError(t, kvDB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> ```

I think you're going to have a bad time if the callback gets called more than once. You should reset the `jobs` slice in the callback. 
___
*[pkg/jobs/registry_test.go, line 350 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcrp8D2uIDrF1mxkbg:-Mfcrp8D2uIDrF1mxkbh:bvqpv4w) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L350)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 
> 				// and description? I was expecting that the type will be "IMPORT" in
> 				// the jobs table when the job is created, but it occurred to be NULL. Is
> 				// that expected? Similarly, description is NULL.
> ```
</details>

When you say `type` what do you mean? Which table? The description is a field in the record so it should be what you set it to. Maybe set it to some expected values. 
___
*[pkg/sql/drop_table.go, line 509 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcsBHtESB5LLN011hP:-MfcsBHtESB5LLN011hQ:b-bykeuw) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/sql/drop_table.go#L509)):*
> ```Go
> 	for _, mj := range tableDesc.MutationJobs {
> 		jobID := jobspb.JobID(mj.JobID)
> 		if record, exists := p.ExtendedEvalContext().SchemaChangeJobCache[tableDesc.ID]; exists {
> ```

This deserves some commentary for future readers as well as some commentary about this behavior on the function itself. 
___
*[pkg/sql/table.go, line 119 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcsmaAETB776m2GlEi:-MfcsmaAETB776m2GlEj:b3dcbo6) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/sql/table.go#L119)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>

> ```
> 	var record *jobs.Record
> 	if cachedRecord, ok := p.extendedEvalCtx.SchemaChangeJobCache[tableDesc.ID]; ok {
> 		record = cachedRecord
> 	}
> ```
</details>

nit: replace this with the following code to allow you to eliminate some code. 

```go
record, recordExists :=  p.extendedEvalCtx.SchemaChangeJobCache[tableDesc.ID]
```


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/bench/rttanalysis/testdata/benchmark_expectations, line 80 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcknDe6Bur1HJ6OGHh:-Mfh7i1O0sBSOgk5u2RZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/bench/rttanalysis/testdata/benchmark_expectations#L80)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 26,Truncate/truncate_1_column_0_rows
> 26,Truncate/truncate_1_column_1_row
> 26,Truncate/truncate_1_column_2_rows
> 26,Truncate/truncate_2_column_0_rows
> 26,Truncate/truncate_2_column_1_rows
> 26,Truncate/truncate_2_column_2_rows
> ```
I have a feeling we're going to want to retain the range here of round-trips here (put the 26-27 back). 
</blockquote></details>

Done.
___
*[pkg/jobs/jobs.go, line 81 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfJAXnHLjDu2QZWm-g1-r2-81:-Mfh8FbS6gPCl9QS3fg8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/jobs/jobs.go#L81)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Yeah, my advice is that we add `ID` as a field to record. It should simplify the interfaces. 
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 375 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfctsoy8qg5KjUp9whS:-MfhAbncDElOCuawTrjt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L375)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> no further jobs are created and the error is returned.

I find this language confusing. The caller shouldn't need to care about other jobs having been created. It makes it sound like it is not a transactional operation. 
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 378 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcwu1wDAApIJ8Yg4cb:-MfhCQm71_vTSQI3lP2W:blq0w31) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L378)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Do callers want the `*Job` or would they be happy with just a slice of IDs? I feel like IDs might be better.
</blockquote></details>

That's right. I have changed it to IDs.
___
*[pkg/jobs/registry.go, line 382 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcu92e1dLlDiHL3z72:-MfhF-GYEDKlyiaHtUHl:b-h45isc) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L382)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you refactor this a bit into a helper to write a batch and a helper to generate the statement and args. I think it should read something like:

```
created := make([]*Job, 0, len(records)
for toCreate := records; len(toCreate) > 0; {
    const maxBatchSize = 100
    batchSize := len(toCreate)
    if batchSize > maxBatchSize {
        batchSize = maxBatchSize
    }
    createdInBatch, err := r.createJobsWithTxnInBatch(ctx, toCreate[:batchSize])
    if err != nil {
        return nil, err
    }
    created = append(created, createdInBatch...)
    toCreate = toCreate[batchSize:]
}

func (r *Registry) createJobsWithTxnInBatch(ctx context.Context, txn *kv.Txn records []*Record) ([]*Job, error) {
    s, err  := r.sqlInstance.Session(ctx)
    if err != nil {
        return nil, errors.Wrap(err, "error getting live session")
    }
    stmt, args := batchJobInsertStmt(s, records)
    if err := r.ex.Exec(ctx, "job-rows-batch-insert", txn, stmt, args...); err != nil {
        return nil, err
    }
    // return the jobs etc. Maybe construct them when you construct the statement and args. 
}
```
</blockquote></details>

This is much better. Thanks for the direction. I have updated.
___
*[pkg/jobs/registry.go, line 454 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcl1tF220t_nj1_YFo:-MfhXJWYDv1cwjlHpASs:bdhsuxe) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry.go#L454)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 			if _, err = r.ex.Exec(ctx, "job-rows-batch-insert", txn, stmtSB.String(), args...,
> 			); err != nil {
> ```
nit: I find this wrapping to be bizarre. My preference is:

```go
		if _, err = r.ex.Exec(
			ctx, "job-rows-batch-insert", txn, stmtSB.String(), args...,
		); err != nil {
```
</blockquote></details>

Thanks, I have updated.
___
*[pkg/jobs/registry_test.go, line 302 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcrZaJ-SOdN04NLpSY:-MfhXmgsDK00lfFSErux:b-g35an3) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L302)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Isn't this covered above?
</blockquote></details>

Good catch, it's by mistake.
___
*[pkg/jobs/registry_test.go, line 334 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcrg254D2oycJJylk0:-MfhtMW_0Ujb_spct6sm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L334)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I think you're going to have a bad time if the callback gets called more than once. You should reset the `jobs` slice in the callback. 
</blockquote></details>

Done.
___
*[pkg/jobs/registry_test.go, line 350 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcrp8D2uIDrF1mxkbg:-Mfhb8VN7oE1VJ5XnArC:blpiz88) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L350)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 
> 				// and description? I was expecting that the type will be "IMPORT" in
> 				// the jobs table when the job is created, but it occurred to be NULL. Is
> 				// that expected? Similarly, description is NULL.
> ```
When you say `type` what do you mean? Which table? The description is a field in the record so it should be what you set it to. Maybe set it to some expected values. 
</blockquote></details>

It was job_type in the internal table as we discussed offline.
___
*[pkg/sql/drop_table.go, line 509 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcsBHtESB5LLN011hP:-MfhtP5t7LLk-6zN0smM:b-k1y5pj) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/sql/drop_table.go#L509)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

This deserves some commentary for future readers as well as some commentary about this behavior on the function itself. 
</blockquote></details>

OK. Done.
___
*[pkg/sql/planner.go, line 96 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfN1ZkROsZJINEQBmxO-r2-96:-MfhqFzK95PhKJ_z6qcJ:bhm7x63) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6c63562b7e43b92a603653f7e29895fb5b97013/pkg/sql/planner.go#L96)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> the test logs show messages related to throttling and ranges going down.

I have a feeling that that's just noise at the end of the test and is not the true failure. 
</blockquote></details>

I have updated the structure.
___
*[pkg/sql/table.go, line 119 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfcsmaAETB776m2GlEi:-Mfhs8f72RIH57mk2O8g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/sql/table.go#L119)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	var record *jobs.Record
> 	if cachedRecord, ok := p.extendedEvalCtx.SchemaChangeJobCache[tableDesc.ID]; ok {
> 		record = cachedRecord
> 	}
> ```
nit: replace this with the following code to allow you to eliminate some code. 

```go
record, recordExists :=  p.extendedEvalCtx.SchemaChangeJobCache[tableDesc.ID]
```

</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
Get the linter passing and fix the nits and I think this is good to go. 

> Reviewed 1 of 3 files at r1, 1 of 14 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-MfiNCwF3PjdPuCUEw_w:b-9idy80)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/jobs/registry_test.go, line 334 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfcrg254D2oycJJylk0:-MfiJiCH7Xg4Fb5QeWYD:bykvkam) ([raw file](https://github.com/cockroachdb/cockroach/blob/622cbc62c33a083ca9b767ff6bd830d7e352662e/pkg/jobs/registry_test.go#L334)):*
<details><summary><i>Previously, sajjadrizvi (Sajjad Rizvi) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not sure what I was talking about here
___
*[pkg/sql/drop_table.go, line 515 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfiKGDb1JuYULKrP5hQ:-MfiKGDb1JuYULKrP5hR:bnz8cic) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/sql/drop_table.go#L515)):*
> ```Go
> ) error {
> 	for _, mj := range tableDesc.MutationJobs {
> 		mutationID := jobspb.JobID(mj.JobID)
> ```

mutationID is confusing as the name of a variable storing a job ID. MutatationID is a differnent concept which is related to jobs. `jobID` was a pretty good name, no?
___
*[pkg/sql/drop_table.go, line 521 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfiL7m6BVwpa5RTuBbJ:-MfiL7m6BVwpa5RTuBbK:b-2273pd) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/sql/drop_table.go#L521)):*
> ```
> 		if record, exists := p.ExtendedEvalContext().SchemaChangeJobCache[tableDesc.ID]; exists {
> 			if record.JobID == mutationID {
> ```
nit: inline the condition with `&&`


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/drop_table.go, line 515 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-MfiKGDb1JuYULKrP5hQ:-MfiU9pH1gcffhjBbcjv:b-1z0l25) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/sql/drop_table.go#L515)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

mutationID is confusing as the name of a variable storing a job ID. MutatationID is a differnent concept which is related to jobs. `jobID` was a pretty good name, no?
</blockquote></details>

Yeah, I didn't realize mutation ID is an existing term.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/bench/rttanalysis/testdata/benchmark_expectations, line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2Mx64aYW1kq6GvmT:-Mfn2Mx64aYW1kq6GvmU:boeycdr) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/bench/rttanalysis/testdata/benchmark_expectations#L74)):*
> ```
> 1,SystemDatabaseQueries/select_system.users_with_empty_database_Name
> 1,SystemDatabaseQueries/select_system.users_with_schema_Name
> 2,SystemDatabaseQueries/select_system.users_without_schema_Name
> ```
What is this about?
___
*[pkg/jobs/registry.go, line 416 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2veO6LrfvzV6T0GC:-Mfn2veO6LrfvzV6T0GD:b-cz0x4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L416)):*
> ```Go
> 	sessionID sqlliveness.SessionID, txn *kv.Txn, records []*Record,
> ) (string, []interface{}, []jobspb.JobID, error) {
> 	const columnValues = " ($%d, $%d, $%d, $%d, $%d, $%d)"
> ```

I don't think we're going to need this with the refactor I'm about to propose. 
___
*[pkg/jobs/registry.go, line 424 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn3Zwh1akSCiaTfRYR:-Mfn3Zwh1akSCiaTfRYS:b-sw8yhk) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L424)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> 	start := timeutil.Now()
> 	if txn != nil {
> 		start = txn.ReadTimestamp().GoTime()
> 	}
> 	modifiedMicros := timeutil.ToUnixMicros(start)
> ```
</details>

nit: lift this above this call and just pass in the timestamp
___
*[pkg/jobs/registry.go, line 427 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2j20DZD6O7wefjr8:-Mfn2j20DZD6O7wefjr9:b-lmf5ou) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L427)):*
> ```Go
> 
> 	// We have six columns per job.
> 	args := make([]interface{}, 0, len(records)*6)
> ```

I'm going to make a nit-picky suggestion for how to make this feel a little cleaner:

```go
instanceID := r.ID()
const  numColumns = 6
columns := [numColumns]string{`id`, `status`, `payload`, `progress`, `claim_session_id`, `claim_instance_id`}
marshalPanic = func(m proto.Message) []byte {
      data, err := protoutil.Marshal(m)
       if err != nil {
           panic(err)
       }
       return data
 }
valueFns := map[string]func(*Record) interface{}{
       `id`: func(r *Record) interface{} { return r.ID },
       `status`: func(r *Record) interface{} { return StatusRunning },
       `payload`: func(r *Record) interface{} { return marshalPanic(r.Payload) },
       `progress`: func(r *Record) interface{} { return marshalPanic(r.Progress) },
       `claim_session_id`: func(r *Record) interface{} { return sessionID.UnsafeBytes() },
       `claim_instance_id`: func(r *Record) interface{} { return instanceID },
 }
appendValues := func(r *Record, val *[]interface{}) (err error) {
        defer func() {
            switch r := recover(); t.(type) {
            case nil:
            case error:
                err = errors.CombineErrors(err, errors.Wrap(r, "encoding job %d", r.ID)
            default:
                panic(r)
            }
        }()
        for _, c := range columns {
            (*vals) = append(*vals, valueFns[c](r)
        }
        return nil
 }
args := make([]interface{}, 0, len(records) * numColumns)
jobIDs := make([]jobspb.JobID, 0, len(records))

var buf strings.Builder
buf.WriteString(`INSERT INTO system.jobs (`
buf.WriteString(strings.Join(columns, ", "))
buf.WriteString(`) VALUES `)
argIdx := 0
for i, r := range records {
    if i > 0 {
        buf.WriteString(", ")
    }
    buf.WriteString("("
    for _, c := range columns {
         buf.WriteString("$")
         buf.WriteString(strconv.Itoa(argIdx))
         argIdx++
    }
    buf.WriteString(")")
    r.Progress.ModifiedMicros = modifiedMicros
    if err := appendValues(r, &args); err != nil {
        return "", nil, nil, err
    }
    jobIDs = append(jobIDs, job.ID)
}
return stmtSB.String(), args, jobIDs, nil
```


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/bench/rttanalysis/testdata/benchmark_expectations, line 74 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2Mx64aYW1kq6GvmT:-MfrdozW0fgWgYw-qYn9:b-hr9008) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/bench/rttanalysis/testdata/benchmark_expectations#L74)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 1,SystemDatabaseQueries/select_system.users_with_empty_database_Name
> 1,SystemDatabaseQueries/select_system.users_with_schema_Name
> 2,SystemDatabaseQueries/select_system.users_without_schema_Name
> ```
What is this about?
</blockquote></details>

I am not sure how it happened. I have now updated only Grant test numbers.
___
*[pkg/jobs/registry.go, line 416 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2veO6LrfvzV6T0GC:-MfrfbUEAqWV3iCm9BsR:b2nhktp) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L416)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I don't think we're going to need this with the refactor I'm about to propose. 
</blockquote></details>

OK.
___
*[pkg/jobs/registry.go, line 424 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn3Zwh1akSCiaTfRYR:-MfrfPI10o-PwRRwa03i:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L424)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	start := timeutil.Now()
> 	if txn != nil {
> 		start = txn.ReadTimestamp().GoTime()
> 	}
> 	modifiedMicros := timeutil.ToUnixMicros(start)
> ```
nit: lift this above this call and just pass in the timestamp
</blockquote></details>

Done.
___
*[pkg/jobs/registry.go, line 427 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mfn2j20DZD6O7wefjr8:-MfshvLL3weUy4fCKvgC:biacqzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/29c3d177c89e704e56c721ae7009f5f8382de3e1/pkg/jobs/registry.go#L427)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm going to make a nit-picky suggestion for how to make this feel a little cleaner:

```go
instanceID := r.ID()
const  numColumns = 6
columns := [numColumns]string{`id`, `status`, `payload`, `progress`, `claim_session_id`, `claim_instance_id`}
marshalPanic = func(m proto.Message) []byte {
      data, err := protoutil.Marshal(m)
       if err != nil {
           panic(err)
       }
       return data
 }
valueFns := map[string]func(*Record) interface{}{
       `id`: func(r *Record) interface{} { return r.ID },
       `status`: func(r *Record) interface{} { return StatusRunning },
       `payload`: func(r *Record) interface{} { return marshalPanic(r.Payload) },
       `progress`: func(r *Record) interface{} { return marshalPanic(r.Progress) },
       `claim_session_id`: func(r *Record) interface{} { return sessionID.UnsafeBytes() },
       `claim_instance_id`: func(r *Record) interface{} { return instanceID },
 }
appendValues := func(r *Record, val *[]interface{}) (err error) {
        defer func() {
            switch r := recover(); t.(type) {
            case nil:
            case error:
                err = errors.CombineErrors(err, errors.Wrap(r, "encoding job %d", r.ID)
            default:
                panic(r)
            }
        }()
        for _, c := range columns {
            (*vals) = append(*vals, valueFns[c](r)
        }
        return nil
 }
args := make([]interface{}, 0, len(records) * numColumns)
jobIDs := make([]jobspb.JobID, 0, len(records))

var buf strings.Builder
buf.WriteString(`INSERT INTO system.jobs (`
buf.WriteString(strings.Join(columns, ", "))
buf.WriteString(`) VALUES `)
argIdx := 0
for i, r := range records {
    if i > 0 {
        buf.WriteString(", ")
    }
    buf.WriteString("("
    for _, c := range columns {
         buf.WriteString("$")
         buf.WriteString(strconv.Itoa(argIdx))
         argIdx++
    }
    buf.WriteString(")")
    r.Progress.ModifiedMicros = modifiedMicros
    if err := appendValues(r, &args); err != nil {
        return "", nil, nil, err
    }
    jobIDs = append(jobIDs, job.ID)
}
return stmtSB.String(), args, jobIDs, nil
```
</blockquote></details>

I have updated accordingly.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> mod some nits and CI

> Reviewed 6 of 14 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-Mg64vhKD4754-iO3pVK:b9an45y)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/bench/rttanalysis/testdata/benchmark_expectations, line 43 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg64pBc6VwiroUOdehp:-Mg64pBc6VwiroUOdehq:b-lnizqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/bench/rttanalysis/testdata/benchmark_expectations#L43)):*
> ```
> 19,DropSequence/drop_1_sequence
> 27,DropSequence/drop_2_sequences
> 35,DropSequence/drop_3_sequences
> ```

seems like some of these have changed. 
___
*[pkg/jobs/registry.go, line 462 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg63cSZ3wggss7QmtLp:-Mg63cSZ3wggss7QmtLq:b2e1q5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/jobs/registry.go#L462)):*
> ```Go
> 	appendValues := func(rec *Record, vals *[]interface{}) (err error) {
> 		defer func() {
> 			switch ret := recover(); ret.(type) {
> ```

`ret` is a confusing name for the recovered value. It's explicitly not something being returned. 
___
*[pkg/sql/conn_executor.go, line 1092 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg640giDZ-WRIrznPPS:-Mg640giDZ-WRIrznPPT:b-29bhfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/sql/conn_executor.go#L1092)):*
> ```Go
> 		jobs jobsCollection
> 
> 		// schemaChangeJobsCache is a map of descriptor IDs to job Records.
> ```

I wonder if this deserves a new name.
___
*[pkg/sql/conn_executor_exec.go, line 809 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg6481s6lsLVxC6mHHm:-Mg6481s6lsLVxC6mHHn:b-n5iulm) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/sql/conn_executor_exec.go#L809)):*
> ```Go
> // createJobs creates jobs for the records cached in schemaChangeJobsCache
> // during this transaction.
> func (ex *connExecutor) createJobs(ctx context.Context) error {
> ```

nit: I'd put this below `commitSQLTransactionInternal`. I usually like to read functions which invoke helpers before the helpers. 
___
*[pkg/sql/planner.go, line 96 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg64IDa8vSbCJMa6iLq:-Mg64IDa8vSbCJMa6iLr:b-i3spwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/sql/planner.go#L96)):*
> ```Go
> 
> 	// SchemaChangeJobCache refers to schemaChangeJobsCache in extraTxnState.
> 	SchemaChangeJobCache map[descpb.ID]*jobs.Record
> ```

In parallel to the other question about names, I wonder if this deserves a new name. `SchemaChangeJobRecords`?


<!-- Sent from Reviewable.io -->

sajjadrizvi reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)

___
*[pkg/sql/planner.go, line 96 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-Mg64IDa8vSbCJMa6iLq:-MgIDpGy3trVAnTUmpyz:b-nj4gau) ([raw file](https://github.com/cockroachdb/cockroach/blob/4e6017eb085557db3d6b8645c5c2816b1c9c55e7/pkg/sql/planner.go#L96)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

In parallel to the other question about names, I wonder if this deserves a new name. `SchemaChangeJobRecords`?
</blockquote></details>

Done, I have updated the name to `SchemaChangeJobRecords`.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, bors it when CI passes. 

> Reviewed 13 of 14 files at r14.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67991#-:-MgLb_gk1Y7oH2G3VZDw:bf7n9u5)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner and @sajjadrizvi)



<!-- Sent from Reviewable.io -->

sajjadrizvi commented:
TFTR!

bors r=ajwerner
